### PR TITLE
fix: updated to handle older git installs

### PR DESCRIPTION
**Pull request by gherlein**
fix: updated to handle older git installs

git before 2.28 did not support "init -b" which causes failures.  Ubuntu 20.04 has an older git and is in widespread use.  This change detects the version of git installed and if it is too old it changes the branch name to "main" manually.  You cannot do this change on an empty repo though, you must do a commit first.  So the sequence of git commands is different for older versions too, doing the add+commit and then changing the default branch to main.

Tested on Ubuntu 20.04 with git version 2.25.1 and again after updating to a new version.  Works in both cases.

No breaking changes

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/cli/cmds/new.ts [hunk 1]
```diff
@@ -244,10 +244,30 @@ async function initProject(baseDir: string, type: inventory.ProjectType, args: a
   }
 
   if (args.git) {
+
     const git = (cmd: string) => exec(`git ${cmd}`, { cwd: baseDir });
-    git('init -b main');
-    git('add .');
-    git('commit --allow-empty -m "chore: project created with projen"');
+    const gitversion = execCapture('git --version', { cwd: baseDir })
+      .toString()
```
Comment (eladb) [inline src/cli/cmds/new.ts:250]: Can we use `semver` library here to parse and compare?

File: src/cli/cmds/new.ts [hunk 2]
```diff
@@ -244,10 +244,30 @@ async function initProject(baseDir: string, type: inventory.ProjectType, args: a
...
+    }
+    if ((major > + 2) && (minor >= 28)) {
+      git('init -b main');
+      git('add .');
+      git('commit --allow-empty -m "chore: project created with projen"');
+    } else {
+      git('init');
+      git('add .');
+      git('commit --allow-empty -m "chore: project created with projen"');
+      console.log('older version of git detected, changing branch name to main');
```
Comment (eladb) [inline src/cli/cmds/new.ts:267]: Use `logger`

**Review by eladb — CHANGES_REQUESTED** [review]

File: src/cli/cmds/new.ts [hunk 3]
```diff
@@ -245,9 +247,23 @@ async function initProject(baseDir: string, type: inventory.ProjectType, args: a
...
+    const gitversion:string = execCapture('git --version', { cwd: baseDir })
+      .toString()
+      .replace(/[a-z,/s]/g, '')
+      .trim();
+    logging.info('system using git version ',gitversion);
+    if ( gitversion && semver.gte(gitversion, '2.28.0') ) {
+      git('init -b main');
+      git('add .');
+      git('commit --allow-empty -m "chore: project created with projen"');
+      logging.info('default branch name set to main');
```
Comment (eladb) [inline src/cli/cmds/new.ts:259]: Either change to debug or omit:

```suggestion
```

**Review by eladb — CHANGES_REQUESTED** [review]

File: src/cli/cmds/new.ts [hunk 4]
```diff
@@ -245,9 +247,23 @@ async function initProject(baseDir: string, type: inventory.ProjectType, args: a
...
+    } else {
+      git('init');
+      git('add .');
+      git('commit --allow-empty -m "chore: project created with projen"');
+      logging.info('older version of git detected, changed default branch name to main');
+      git('branch -m main');
```
[the remark this answers is not in the document]
Comment (gherlein) [reply src/cli/cmds/new.ts]: yes, good idea - the --force will be better

**Comment by eladb** [comment]
## Please read this to avoid merge hell

Hola! apologies for hassle. I've enabled Prettify as part of #1453. This means that most of the source files have changed and you will likely see quite a lot of merge conflicts when you try to merge from `main`.

*But there's an easy way to handle this.*

First, merge your branch from this tag:

```sh
$ git fetch origin
$ git merge 1020d0ad3f95bb4e1201f8640d440ffb43ccf230
```

Edit the `.projenrc.js` file in your branch and add this to **line 57** (right
after `codeCov: true`):

```js
  prettier: true,
```

Now, run:

```sh
$ git clean -fdx .
$ yarn
$ bash projen.bash
$ yarn eslint
$ git add .
$ git commit -m 'format code'
```

This will reformat all the code in your branch to match `main`.

Now, merge from main:

```sh
$ git merge origin/main
```

Sorry again for the inconvenience...

**Comment by eladb** [comment]
I _think_ that you just need to now run:

```sh
git merge origin/main
```

To sync your branch from main.

**Comment by gherlein** [comment]
***@***.***:~/src/constructs/after/projen$ git merge
origin/main
Auto-merging yarn.lock
Auto-merging src/typescript/typescript.ts
CONFLICT (content): Merge conflict in src/typescript/typescript.ts
Auto-merging src/javascript/jest.ts
CONFLICT (content): Merge conflict in src/javascript/jest.ts
Auto-merging src/javascript/eslint.ts
Auto-merging src/index.ts
CONFLICT (content): Merge conflict in src/index.ts
Auto-merging src/cli/cmds/new.ts
CONFLICT (content): Merge conflict in src/cli/cmds/new.ts
Auto-merging src/build/build-workflow.ts
CONFLICT (content): Merge conflict in src/build/build-workflow.ts
Auto-merging package.json
CONFLICT (content): Merge conflict in package.json
Auto-merging .projenrc.js
CONFLICT (content): Merge conflict in .projenrc.js
Auto-merging .projen/tasks.json
Auto-merging .projen/files.json
Auto-merging .gitignore
Removing .github/workflows/integ.yml
Auto-merging .gitattributes
Automatic merge failed; fix conflicts and then commit the result.

Are those expected?  Should I just go fix them?

Starting over is seeming way faster...


--------------------------------------------------------------------------------------------------------------
Greg Herlein   |   +1-415-368-7546   |   LinkedIn
<http://www.linkedin.com/in/gherlein>

**Comment by eladb** [comment]
Looks like a reasonable set of conflicts. But if you feel more comfortable starting over, by all means

On Jan 3, 2022, at 19:37, Greg Herlein ***@***.***> wrote:

﻿

***@***.***:~/src/constructs/after/projen$ git merge
origin/main
Auto-merging yarn.lock
Auto-merging src/typescript/typescript.ts
CONFLICT (content): Merge conflict in src/typescript/typescript.ts
Auto-merging src/javascript/jest.ts
CONFLICT (content): Merge conflict in src/javascript/jest.ts
Auto-merging src/javascript/eslint.ts
Auto-merging src/index.ts
CONFLICT (content): Merge conflict in src/index.ts
Auto-merging src/cli/cmds/new.ts
CONFLICT (content): Merge conflict in src/cli/cmds/new.ts
Auto-merging src/build/build-workflow.ts
CONFLICT (content): Merge conflict in src/build/build-workflow.ts
Auto-merging package.json
CONFLICT (content): Merge conflict in package.json
Auto-merging .projenrc.js
CONFLICT (content): Merge conflict in .projenrc.js
Auto-merging .projen/tasks.json
Auto-merging .projen/files.json
Auto-merging .gitignore
Removing .github/workflows/integ.yml
Auto-merging .gitattributes
Automatic merge failed; fix conflicts and then commit the result.

Are those expected? Should I just go fix them?

Starting over is seeming way faster...


--------------------------------------------------------------------------------------------------------------
Greg Herlein | +1-415-368-7546 | LinkedIn
<http://www.linkedin.com/in/gherlein>

**Comment by Chriscbr** [comment]
Superseded by https://github.com/projen/projen/pull/1485

**Pull request closed without merging**